### ISSUE TITLE
cv2.cuda_GpuMat()

### ISSUE THREAD

**Issue by SwHaraday** [ultralytics/yolov5#13723]
### Search before asking

- [x] I checked the docs, issues, and discussions and could not find an answer.

### Question

Hi Glenn,

It's been a while. Thanks to your great work, our yolov5 system has been deployed across 7 lines.

I'm wondering if it is possible that resize and letterbox are done on GPU using cv2.cuda_GpuMat().
I knew these 2 proccesses take 5 to 7 milli-second which is about half of total inference time.
Is it possible to feed cv2.cuda_GpuMat() directly to yolov5s(ver.6.0) model ?

If this is already answered, please kindly guide me.

Thank you for your continuing effort for democratization of object detection.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SwHaraday, thank you for your interest in YOLOv5 🚀 and for sharing that your system is already deployed across 7 lines — that’s awesome to hear 😄
This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist you soon.
Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).
If this is a 🐛 Bug Report, please provide a **minimum reproducible example** to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
For performance and preprocessing questions like this one, it would also help if you could share:
- Your exact YOLOv5 version and install method
- Your OpenCV, Python, PyTorch, and CUDA versions
- A small reproducible code snippet showing your current preprocessing and inference pipeline
- Timing breakdowns for resize, letterbox, host-to-device transfer, and inference
- Whether you are using [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) directly or a custom pipeline
## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```
## Environments
YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by pderrenger** [comment]
Thanks for the kind words, and credit to the YOLO community as well. YOLOv5 v6.0 does not natively accept `cv2.cuda_GpuMat()` inputs, since the default `detect.py` path does `resize`/`letterbox` on CPU and the model expects a `torch.Tensor`; if you want GPU preprocessing, you’ll need a custom wrapper that converts the frame into a CUDA tensor before inference, and it’s also worth checking whether the same bottleneck remains on the latest YOLOv5 code.

**Comment by SwHaraday** [comment]
Dear pderrenger,

Thank you very much for detailed explanation.
I've confirmed resize/letterbox can be done on GPU.
I started reading class DetectMultiBackend in models/common.py. 
I'll find the way to convert cv2.cuda_GpuMat() to torch.Tensor on GPU.

BR Yuji

**Comment by pderrenger** [comment]
Thanks for the update, Yuji — that’s the right direction: `DetectMultiBackend` can take a `torch.Tensor`, so the clean path is usually to keep your GPU preprocessing in a custom loader/wrapper and pass a CUDA BCHW tensor with the expected dtype/normalization into inference; if you run into a specific conversion or shape issue, feel free to share that snippet here and we can take a look.

**Comment by SwHaraday** [comment]
Hi,
I'm retiring.
I thank this TEAM for continuing efforts to democratize vision AI.
YOU people are the BEST !

**Comment by pderrenger** [comment]
Congratulations on your retirement, Yuji, and thank you for the kind words and for building with YOLOv5 over the years; the credit really goes to the whole open-source community, and we wish you all the best in your next chapter.
